Your task is to manually extract key financial data for the last three fiscal years from the 10-K filings of Microsoft, Tesla, and Apple. \
Following the data collection, you will use Python to analyze this data, focusing on trends and insights that could inform the development of an AI-powered financial chatbot.

# Step 1: Data extraction

Navigate to the SEC's EDGAR database:

Microsoft
Tesla
Apple

\
Manual extraction:

For each company, find the 10-K filings for the last three fiscal years.
Extract the following financial figures: Total Revenue, Net Income, Total Assets, Total Liabilities, and Cash Flow from Operating Activities.

\
Organize Your Data:

Compile the extracted data into an Excel spreadsheet for easy reference during your Python analysis.

# Step 2: Preparing your Jupyter Notebook environment

We will be using Google Colab, which is equivalent to Jupyter Notebook

# Step 3: Python analysis in Jupyter

## Load and clean dataset

In [7]:
import pandas as pd

In [8]:
df = pd.read_csv('10-k-fillings - Sheet1.csv')

In [9]:
# View the Dataframe
df

Company  Fiscal Year Total Revenue Net Income Total Assets  \
0  Microsoft         2024       245,122     88,136      512,163   
1  Microsoft         2023       211,915     72,361      411,976   
2  Microsoft         2022       198,270     72,738      364,840   
3      Tesla         2024        97,690      7,153      122,070   
4      Tesla         2023        96,773     14,974      106,618   
5      Tesla         2022        81,462     12,587       82,338   
6      Apple         2024       391,035     93,736      364,980   
7      Apple         2023       383,285     96,995      352,583   
8      Apple         2022       394,328     99,803      352,755   

  Total Liabilities Cash Flow From Ops  
0           243,686            118,548  
1           205,753             87,582  
2           198,298             89,035  
3            48,390             14,923  
4            43,009             13,256  
5            36,440             14,724  
6           308,030            118,254  
7           290,437            110,543  
8           302,083            122,151

In [10]:
# Ensure all elements are numeric value, expect Company name
for col in df.columns:
  if col != 'Company':
    # Remove dollar signs, commas, or other formatting if needed
    df[col] = df[col].replace('[\$,]', '', regex=True)
    # Convert to numeric object
    df[col] = pd.to_numeric(df[col], errors='coerce')

In [11]:
# Check the types of data in the Dataframe
df.dtypes

Company               object
Fiscal Year            int64
Total Revenue          int64
Net Income             int64
Total Assets           int64
Total Liabilities      int64
Cash Flow From Ops     int64
dtype: object

In [12]:
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0  Microsoft         2024         245122       88136        512163   
1  Microsoft         2023         211915       72361        411976   
2  Microsoft         2022         198270       72738        364840   
3      Tesla         2024          97690        7153        122070   
4      Tesla         2023          96773       14974        106618   
5      Tesla         2022          81462       12587         82338   
6      Apple         2024         391035       93736        364980   
7      Apple         2023         383285       96995        352583   
8      Apple         2022         394328       99803        352755   

   Total Liabilities  Cash Flow From Ops  
0             243686              118548  
1             205753               87582  
2             198298               89035  
3              48390               14923  
4              43009               13256  
5              36440               14724  
6             308030              118254  
7             290437              110543  
8             302083              122151

## Analyze data

Use pandas to calculate year-over-year changes for each financial metric.

You can do this by creating new columns in your DataFrame that represent the percentage change from one year to the next.

In [13]:
# Ensure 'Fiscal Year' is sorted in ascending order within each Company
df = df.sort_values(by=['Company', 'Fiscal Year'])
# Reset all indexes
df = df.reset_index(drop=True)
df.head()

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2022         394328       99803        352755   
1      Apple         2023         383285       96995        352583   
2      Apple         2024         391035       93736        364980   
3  Microsoft         2022         198270       72738        364840   
4  Microsoft         2023         211915       72361        411976   

   Total Liabilities  Cash Flow From Ops  
0             302083              122151  
1             290437              110543  
2             308030              118254  
3             198298               89035  
4             205753               87582

In [14]:
# # Calculate year-over-year growth rates for Total Revenue and Net Income
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [15]:
# Fill NA values that result from pct_change calculations with 0 or an appropriate value
df.fillna(0, inplace=True)

In [16]:
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2022         394328       99803        352755   
1      Apple         2023         383285       96995        352583   
2      Apple         2024         391035       93736        364980   
3  Microsoft         2022         198270       72738        364840   
4  Microsoft         2023         211915       72361        411976   
5  Microsoft         2024         245122       88136        512163   
6      Tesla         2022          81462       12587         82338   
7      Tesla         2023          96773       14974        106618   
8      Tesla         2024          97690        7153        122070   

   Total Liabilities  Cash Flow From Ops  Revenue Growth (%)  \
0             302083              122151            0.000000   
1             290437              110543           -2.800461   
2             308030              118254            2.021994   
3             198298               89035            0.000000   
4             205753               87582            6.882030   
5             243686              118548           15.669962   
6              36440               14724            0.000000   
7              43009               13256           18.795267   
8              48390               14923            0.947578   

   Net Income Growth (%)  
0               0.000000  
1              -2.813543  
2              -3.359967  
3               0.000000  
4              -0.518299  
5              21.800417  
6               0.000000  
7              18.964010  
8             -52.230533

In [17]:
# Export dataframe as CSV file
df.to_csv('10-k-fillings-update.csv', index=False)

In [18]:
# Build a summary of findings
# Company's average performance for past three years
summary = df.groupby('Company').agg({
  'Total Revenue': 'mean',
  'Net Income': 'mean',
  'Revenue Growth (%)': 'mean',
  'Net Income Growth (%)': 'mean'
}).reset_index()
# Round up to 3 digits
summary = summary.round(3)

print("\nCompany Financial Performance from fiscal year 2022 to 2024:")
print(summary)


Company Financial Performance from fiscal year 2022 to 2024:
     Company  Total Revenue  Net Income  Revenue Growth (%)  \
0      Apple     389549.333   96844.667              -0.259   
1  Microsoft     218435.667   77745.000               7.517   
2      Tesla      91975.000   11571.333               6.581   

   Net Income Growth (%)  
0                 -2.058  
1                  7.094  
2                -11.089  


## Conclusion (Generated by Chatgpt based on data):

* Microsoft:
  * Strong growth in revenue (+15.67%) and net income (+21.8%) from 2023 to 2024.

  * Revenue grew steadily across all years.

  * Indicates healthy operational and profit expansion.

* Apple:
  * Slight revenue decline in 2023 (−2.80%), followed by a mild rebound in 2024 (+2.02%).

  * Net income consistently declined, with a −3.36% drop in 2024.

  * While revenue is stabilizing, profits are slipping — possibly due to cost pressures or margin erosion.

* Tesla:
  * Revenue growth slowed dramatically in 2024 (+0.95% vs. +18.80% in 2023).

  * Net income plunged by −52.23% in 2024 despite modest revenue growth.

  * Suggests serious margin compression or rising costs — a potential red flag.

# Step 4: Documentation and submission

* Document your analysis: Use the markdown feature in Jupyter Notebook to document your methodology, observations, and conclusions throughout the notebook.
* Export your notebook: Once your analysis is complete, export your Jupyter Notebook as a PDF or HTML file for submission.
 * You can do this from the "File" menu in Jupyter, selecting "Download as" and then choosing your preferred format.